### PR TITLE
Add github action to sync api docs to the user docs site

### PR DESCRIPTION
**Pull request by tinygrasshopper**
- Currently this generates references docs, based on the tags supplied in the open api docs

- Will add altering if the job fails as a separate PR

### CODE REVIEW

**Review by jgresty — COMMENTED** [review]
Couple of things:
 - There is no linting or build step so we wont know if anything is broken until we try to run it next
 - Documentation could be far better. Even a basic what this does and why in a readme would be appreciated
 - There is a lot of behaviour here that is untested. I would not be comfortable updating dependencies etc

**Comment by jgresty** [comment]
Oh and alerting, how will we know if this fails?

File: tools/api-docs-generator/spec_fetcher_test.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	}{
+		{
+			name:     "gets the latest version",
+			versions: []string{"2024-03-12~experimental", "2024-03-12~beta", "2024-03-12", "2024-03-15~experimental", "2024-03-15~beta", "2024-03-15", "2024-04-11~experimental", "2024-04-11~beta", "2024-04-11", "2024-04-22~experimental", "2024-04-22~beta", "2024-04-22", "2024-04-25~experimental", "2024-04-25~beta", "2024-04-25", "2024-04-29~experimental", "2024-04-29~beta", "2024-04-29", "2024-05-08~experimental", "2024-05-08~beta", "2024-05-08"},
+			want:     "2024-05-08",
+		},
+		{
+			name:     "gets the latest GA version",
+			versions: []string{"2024-03-12~experimental", "2024-03-12~beta", "2024-03-12", "2024-03-15~experimental", "2024-03-15~beta", "2024-03-15", "2024-04-11~experimental", "2024-04-11~beta", "2024-04-11", "2024-04-22~experimental", "2024-04-22~beta", "2024-04-22", "2024-04-25~experimental", "2024-04-25~beta", "2024-04-25", "2024-04-29~experimental", "2024-04-29~beta", "2024-04-29", "2024-05-08~experimental", "2024-05-08~beta"},
+			want:     "2024-04-29",
```
[the remark this answers is not in the document]
Comment (tinygrasshopper) [reply tools/api-docs-generator/spec_fetcher_test.go:19]: `gets the latest GA version`
tests if a non GA version is after the GA version, the GA version is returned

File: tools/api-docs-generator/reference_docs.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+						docsHint:  spec.DocsHint,
+					})
+				}
+			}
+		}
+	}
+
+	var summary []string
+	for label, operation := range aggregatedDocs {
+		if label == "OpenAPI" {
```
[the remark this answers is not in the document]
Comment (tinygrasshopper) [reply tools/api-docs-generator/reference_docs.go:51]: its not a useful domain concept in my mind, but fair enough, it should be added as reference

File: tools/api-docs-generator/reference_docs.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		if err != nil {
+			return err
+		}
+		summary = append(summary, fmt.Sprintf("* [%s](%s)\n", label, path.Join(config.Output.APIReferencePath, labelToFileName(label))))
+
+		fmt.Fprintf(docsFile, `# %s
+
+{%% hint style="info" %%}
+%s
+{%% endhint %%}`, label, operation[0].docsHint)
```
[the remark this answers is not in the document]
Comment (tinygrasshopper) [reply tools/api-docs-generator/reference_docs.go:65]: all docs hints are the same for a given tag

File: tools/api-docs-generator/reference_docs.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jgresty) [reply tools/api-docs-generator/reference_docs.go:51]: I agree that we don't need it, what I meant here was that we are building the `aggregatedDocs` slice just above here then not using it again. We could do this test when building the slice (line ~36) instead and skipping over them there.

File: tools/api-docs-generator/reference_docs.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgresty) [reply tools/api-docs-generator/reference_docs.go:65]: I see it comes from `spec`. This is the wrong data structure given that but that is a trivial matter.

However that does bring up the question if there are overlapping paths in the specs, I don't think that is the case for our v1 and rest but a potential footgun that should be called out somewhere (even if just a comment)

File: tools/api-docs-generator/spec_fetcher_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jgresty) [reply tools/api-docs-generator/spec_fetcher_test.go:19]: Then `ignores later non-GA versions` may be a better name

File: tools/api-docs-generator/reference_docs.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tinygrasshopper) [reply tools/api-docs-generator/reference_docs.go]: I dont think we can have overlapping paths in spec, the routing wont work; v1/vrest should be unique

File: .github/workflows/test-generator.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+      - '.github/workflows/test-generator.yml'
+
+jobs:
+  test:
+    name: test-api-docs-generator
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          #TODO: Remove before merge
```
Comment (jgresty) [inline .github/workflows/test-generator.yml:17]: Another todo

File: tools/api-docs-generator/Makefile [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+	@go run . config.yml ../..
+
+lint: tidy
+	go run github.com/golangci/golangci-lint/cmd/golangci-lint run
+
+format: tidy
+	go run golang.org/x/tools/cmd/goimports -w -local=github.com/snyk/user-docs/tools/api-docs-generator .
+
+tidy:
+	go mod tidy
```
Comment (jgresty) [inline tools/api-docs-generator/Makefile:14]: nit: all of these should be marked `.PHONY` as they do not output the target

File: tools/api-docs-generator/Makefile [hunk 6]
```diff
@@ -0,0 +1,14 @@
+test: tidy lint
+	go test ./...
+
+run:
+	@go run . config.yml ../..
+
+lint: tidy
+	go run github.com/golangci/golangci-lint/cmd/golangci-lint run
```
Comment (jgresty) [inline tools/api-docs-generator/Makefile:8]: Please pin the version, there are occasionally breaking changes

File: tools/api-docs-generator/fetcher/spec_fetcher.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	defer resp.Body.Close()
+
+	jsonSpec, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	formattedSpec := bytes.NewBufferString("")
+	err = json.Indent(formattedSpec, jsonSpec, "", "  ")
```
Comment (jgresty) [inline tools/api-docs-generator/fetcher/spec_fetcher.go:56]: json keys are unordered, I'm not sure if vervet generation is stable if content is the same but something else changes that is unrelated to the current spec - eg a new version is published

File: tools/api-docs-generator/fetcher/spec_fetcher.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+		return err
+	}
+
+	return os.WriteFile(path.Join(directory, cfg.Fetcher.Destination), formattedSpec.Bytes(), 0o644)
+}
+
+func getLatestGAVersion(versions []string) string {
+	gaVersions := []string{}
+	for _, version := range versions {
+		if !strings.Contains(version, "~") {
```
Comment (jgresty) [inline tools/api-docs-generator/fetcher/spec_fetcher.go:67]: `~ga` is an optional valid suffix for ga versions

File: tools/api-docs-generator/fetcher/spec_fetcher.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+			gaVersions = append(gaVersions, version)
+		}
+	}
+	sort.Strings(gaVersions)
```
Comment (jgresty) [inline tools/api-docs-generator/fetcher/spec_fetcher.go:71]: You're already iterating over the entire list above, seems very inefficient to do that and then sort them after to select one element when you can select the largest element in one pass

File: .github/workflows/sync-api-docs.yml [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+          exit $result_code
+      - uses: peter-evans/create-pull-request@v6
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          commit-message: "docs: synchronizing api spec with user-docs"
+          title: "Generate API docs from spec"
+          body: |
+            This PR was automatically generated by the API docs synchronization action. Please review the changes and merge if they look good.
+            ```
+            ${{ steps.generate.outputs.MENU }}
```
Comment (jgresty) [inline .github/workflows/sync-api-docs.yml:42]: I think we need to have a bit more explanation about what to do with this menu in the pull request. At the moment it is tribal knowledge that can easily be misunderstood.

**Review by jgresty — APPROVED** [review]
This is looking good!

I'm still concerned what will happen if this script runs before a previous pull request is merged, looks like it is using the same branch name every time so will it error? Stack empty commits? No-op? (hopefully the last one)

I can see there were a lot of notes that were previously on Apiary that have been lost on the new V1 docs. Eg: https://snyk.docs.apiary.io/#reference/import-projects . Some of those notes may be redundant but we are losing information in this new system. However the one for licenses is present: https://snyk.docs.apiary.io/#reference/licenses . Was that one manually added? Will this script overwrite it?

File: tools/api-docs-generator/Makefile (hunk 6 above)
Comment (tinygrasshopper) [reply tools/api-docs-generator/Makefile:8]: version is pinned in the go.mod file

File: .github/workflows/sync-api-docs.yml [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+      - name: generate
+        id: generate
+        working-directory: ./tools/api-docs-generator
+        run: |
+          make run | tee /tmp/run.log
+          result_code=${PIPESTATUS[0]}
+          echo 'MENU<<EOF' >> $GITHUB_OUTPUT
+          cat /tmp/run.log >> $GITHUB_OUTPUT
+          echo 'EOF' >> $GITHUB_OUTPUT
+          exit $result_code
```
[the remark this answers is not in the document]
Comment (tinygrasshopper) [reply .github/workflows/sync-api-docs.yml:33]:
```
echo 'MENU<<EOF' >> $GITHUB_OUTPUT
make run >> $GITHUB_OUTPUT
echo 'EOF' >> $GITHUB_OUTPUT
```
Looses the exit status of the run; it will return the exit status of echo

File: tools/api-docs-generator/fetcher/spec_fetcher.go (hunk 8 above)
Comment (tinygrasshopper) [reply tools/api-docs-generator/fetcher/spec_fetcher.go:67]: the suffix is used for version resolution right? When listing the versions, I dont see us rendering ~ga at the end of versions

File: tools/api-docs-generator/fetcher/spec_fetcher.go (hunk 7 above)
Comment (tinygrasshopper) [reply tools/api-docs-generator/fetcher/spec_fetcher.go:56]: interesting point, have not seen this occur yet, keep on eye on the PR if these PRs are generated

**Comment by tinygrasshopper** [comment]
> I can see there were a lot of notes that were previously on Apiary that have been lost on the new V1 docs.
Yep, will add that feature as a followup PR

> I'm still concerned what will happen if this script runs before a previous pull request is merged
The default behaviour of the PR action is to merge it into the previous PR

**Review by mikeromard — APPROVED** [review]
Approving on behalf of docs team.
